share/walrus.1
.\" Man page generated from reStructuredText.
.
.TH WALRUS 1 "April 02, 2021" "v0.1.5rc1" ""
.SH NAME
walrus \- back-port compiler for Python 3.8 assignment expression
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
walrus [\fIoptions\fP] <\fIPython source files and directories...\fP>
.SH DESCRIPTION
.sp
Since PEP 572, Python introduced \fIassignment expressions\fP syntax in
version __3.8__. For those who wish to use \fIassignment expressions\fP
in their code, \fBwalrus\fP provides an intelligent, yet imperfect,
solution of a \fBbackport compiler\fP by replacing \fIassignment expressions\fP
syntax with old\-fashioned syntax, which guarantees you to always write
\fIassignment expressions\fP in Python 3.8 flavour then compile for
compatibility later.
.sp
This man page mainly introduces the CLI options of the \fBwalrus\fP program.
You can also checkout the online documentation at
\fI\%https://bpc\-walrus.readthedocs.io/\fP for more details.
.SH OPTIONS
.SS positional arguments
.INDENT 0.0
.TP
.B SOURCE
Python source files and directories to be converted
.UNINDENT
.SS optional arguments
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-V\fP,\fB  \-\-version
show program\(aqs version number and exit
.TP
.B \-q\fP,\fB  \-\-quiet
run in quiet mode
.UNINDENT
.INDENT 0.0
.TP
.B \-C \fIN\fP, \-\-concurrency \fIN\fP
the number of concurrent processes for conversion
.UNINDENT
.INDENT 0.0
.TP
.B \-\-dry\-run
list the files to be converted without actually performing conversion and archiving
.UNINDENT
.INDENT 0.0
.TP
.B \-s \fI[FILE]\fP, \-\-simple \fI[FILE]\fP
this option tells the program to operate in "simple mode"; if a file name is provided, the program will convert
the file but print conversion result to standard output instead of overwriting the file; if no file names are
provided, read code for conversion from standard input and print conversion result to standard output; in
"simple mode", no file names shall be provided via positional arguments
.UNINDENT
.SS archive options
.sp
backup original files in case there\(aqre any issues
.INDENT 0.0
.TP
.BI \-n\fB a\fR,\fB \ \-\-no\-archive
do not archive original files
.UNINDENT
.INDENT 0.0
.TP
.B \-k \fIPATH\fP, \-\-archive\-path \fIPATH\fP
path to archive original files
.TP
.B \-r \fIARCHIVE_FILE\fP, \-\-recover \fIARCHIVE_FILE\fP
recover files from a given archive file
.UNINDENT
.sp
\-r2                     remove the archive file after recovery
\-r3                     remove the archive file after recovery, and remove the archive directory if it becomes empty
.SS convert options
.sp
conversion configuration
.INDENT 0.0
.TP
.B \-vs \fIVERSION\fP, \-\-vf \fIVERSION\fP, \-\-source\-version \fIVERSION\fP, \-\-from\-version \fIVERSION\fP
parse source code as this Python version
.TP
.B \-l \fILINESEP\fP, \-\-linesep \fILINESEP\fP
line separator (\fBLF\fP, \fBCRLF\fP, \fBCR\fP) to read source files
.TP
.B \-t \fIINDENT\fP, \-\-indentation \fIINDENT\fP
code indentation style, specify an integer for the number of spaces, or \fB\(aqt\(aq\fP/\fB\(aqtab\(aq\fP for tabs
.UNINDENT
.sp
\-n8, \-\-no\-pep8          do not make code insertion \fBPEP 8\fP compliant
.SH ENVIRONMENT
.sp
\fBwalrus\fP currently supports these environment variables:
.INDENT 0.0
.TP
.B WALRUS_QUIET
run in quiet mode
.TP
.B WALRUS_CONCURRENCY
the number of concurrent processes for conversion
.TP
.B WALRUS_DO_ARCHIVE
whether to perform archiving
.TP
.B WALRUS_ARCHIVE_PATH
path to archive original files
.TP
.B WALRUS_SOURCE_VERSION
parse source code as this Python version
.TP
.B WALRUS_LINESEP
line separator to read source files
.TP
.B WALRUS_INDENTATION
code indentation style
.TP
.B WALRUS_PEP8
whether to make code insertion \fBPEP 8\fP compliant
.UNINDENT
.SH SEE ALSO
.sp
pybpc(1), f2format(1), poseur(1), vermin(1)
.SH AUTHOR
Contributors of the Python Backport Compiler project.
See https://github.com/pybpc
.SH COPYRIGHT
walrus is licensed under the MIT License.
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log

## [v0.1.3](https://github.com/pybpc/walrus/tree/v0.1.3) (2019-12-11)
[Full Changelog](https://github.com/pybpc/walrus/compare/v0.1.2.post1...v0.1.3)

## [v0.1.2.post1](https://github.com/pybpc/walrus/tree/v0.1.2.post1) (2019-11-28)
[Full Changelog](https://github.com/pybpc/walrus/compare/v0.1.2...v0.1.2.post1)

## [v0.1.2](https://github.com/pybpc/walrus/tree/v0.1.2) (2019-11-27)
[Full Changelog](https://github.com/pybpc/walrus/compare/v0.1.1...v0.1.2)

## [v0.1.1](https://github.com/pybpc/walrus/tree/v0.1.1) (2019-11-26)
[Full Changelog](https://github.com/pybpc/walrus/compare/v0.1.0...v0.1.1)

## [v0.1.0](https://github.com/pybpc/walrus/tree/v0.1.0) (2019-11-25)
**Fixed bugs:**

- Docstring Injection Vulnerability [\#2](https://github.com/pybpc/walrus/issues/2)

**Merged pull requests:**

- Add testcases [\#1](https://github.com/pybpc/walrus/pull/1) ([gousaiyang](https://github.com/gousaiyang))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


walrus.bak.py
# -*- coding: utf-8 -*-

import argparse
import glob
import locale
import os
import re
import shutil
import sys
import uuid

import parso
import tbtrim

__all__ = ['walrus', 'convert']

# multiprocessing may not be supported
try:        # try first
    import multiprocessing
except ImportError:  # pragma: no cover
    multiprocessing = None
else:       # CPU number if multiprocessing supported
    if os.name == 'posix' and 'SC_NPROCESSORS_CONF' in os.sysconf_names:  # pragma: no cover
        CPU_CNT = os.sysconf('SC_NPROCESSORS_CONF')
    elif hasattr(os, 'sched_getaffinity'):  # pragma: no cover
        CPU_CNT = len(os.sched_getaffinity(0))  # pylint: disable=E1101
    else:  # pragma: no cover
        CPU_CNT = os.cpu_count() or 1
finally:    # alias and aftermath
    mp = multiprocessing
    del multiprocessing

# version string
__version__ = '0.1.0.dev0'

# from configparser
BOOLEAN_STATES = {'1': True, '0': False,
                  'yes': True, 'no': False,
                  'true': True, 'false': False,
                  'on': True, 'off': False}

# environs
LOCALE_ENCODING = locale.getpreferredencoding(False)

# macros
grammar_regex = re.compile(r"grammar(\d)(\d)\.txt")
WALRUS_VERSION = sorted(filter(lambda version: version >= '3.8',  # when Python starts to have walrus operator
                               map(lambda path: '%s.%s' % grammar_regex.match(os.path.split(path)[1]).groups(),
                                   glob.glob(os.path.join(parso.__path__[0], 'python', 'grammar??.txt')))))
del grammar_regex


class ConvertError(SyntaxError):
    """Parso syntax error."""


class ContextError(RuntimeError):
    """Missing conversion context."""


class EnvironError(EnvironmentError):
    """Invalid environment."""


###############################################################################
# Traceback trim (tbtrim)

# root path
ROOT = os.path.dirname(os.path.realpath(__file__))


def predicate(filename):  # pragma: no cover
    if os.path.basename(filename) == 'walrus':
        return True
    return ROOT in os.path.realpath(filename)


tbtrim.set_trim_rule(predicate, strict=True, target=(ConvertError, ContextError))

###############################################################################
# Main convertion implementation

# walrus wrapper template
FUNC_TEMPLATE = '''
def __walrus_wrapper_%(name)s_%(uuid)s():
    """Wrapper function for assignment expression `%(expr)s`."""
    %(keyword)s %(name)s
    %(name)s = %(expr)s
    return %(name)s
'''.splitlines()


def parse(string, source, error_recovery=False):
    """Parse source string.

    Args:
     - `string` -- `str`, context to be converted
     - `source` -- `str`, source of the context
     - `error_recovery` -- `bool`, see `parso.Grammar.parse`

    Envs:
     - `WALRUS_VERSION` -- convert against Python version (same as `--python` option in CLI)

    Returns:
     - `parso.python.tree.Module` -- parso AST

    Raises:
     - `ConvertError` -- when `parso.ParserSyntaxError` raised

    """
    try:
        return parso.parse(string, error_recovery=error_recovery,
                           version=os.getenv('WALRUS_VERSION', WALRUS_VERSION[-1]))
    except parso.ParserSyntaxError as error:
        message = '%s: <%s: %r> from %s' % (error.message, error.error_leaf.token_type,
                                            error.error_leaf.value, source)
        raise ConvertError(message).with_traceback(error.__traceback__) from None


class ConvertContext:
    """Conversion context."""

    @property
    def tabsize(self):
        """Indentation tab size.

        Returns:
         - `int` -- inndentation tab size

        """
        return self._indent

    @property
    def linesep(self):
        """Line separator.

        Returns:
         - `str` -- line separator

        """
        return self._linesep

    @property
    def column(self):
        """Current indentation.

        Returns:
         - `int` -- current inndentation

        """
        return self._column

    def __init__(self, node, indent=0, tabsize=None):
        """"Hold process context.

        Args:
         - `node` -- `Union[parso.python.tree.PythonNode, parso.python.tree.PythonLeaf]`, parso AST
         - `indent` -- `int`, tab size addon for suites
         - `tabsize` -- `int`, indentation tab size

        Envs:
         - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)

        """
        self._linesep = guess_linesep(node)
        self._column = node.get_first_leaf().column + indent
        self._indent = tabsize or indent
        self._func = list()
        self._vars = list()

    def extract(self, node):
        """Process assignment expression (`namedexpr_test`).

        Args:
        - `node` -- `parso.python.tree.PythonNode`, assignment expression node

        Returns:
         - `str` -- converted string

        """
        # split assignment expression
        node_name, _, node_expr = node.children
        name = node_name.value
        expr = walk(node_expr, self)
        uid = uuid.uuid4().hex

        function = self.make_func(name, expr, uid)
        string = '__walrus_wrapper_%s_%s()' % (name, uid)

        if name not in self._vars:
            variable = '%s%s = locals().get(%r)%s' % ('\t'.expandtabs(self.column),
                                                      name, name, self.linesep)
            self._vars.append(variable)
        self._func.append(function)

        return string

    def make_func(self, name, expr, uid):
        """Generate wrapper function.

        Args:
         - `name` -- `str`, variable name
         - `expr` -- `str`, variable expression
         - `uid` -- `str`, hash ID of wrapped node

        Returns:
         - `str` -- wrapper function

        """
        return (
            '%s%s' % (self.linesep, '\t'.expandtabs(self.column))
        ).join(FUNC_TEMPLATE) % dict(keyword='nonlocal' if self.column else 'global',
                                     name=name, expr=expr.strip(), uuid=uid) + self.linesep

    def finalize(self, prefix, suffix):
        """Format final results.

        Args:
         - `prefix` -- `str`, prefix string
         - `suffix` -- `str`, suffix string

        Returns:
         - `str` -- finalised converted string

        """
        variables = self.linesep.join(self._vars)
        functions = self.linesep.join(self._func)

        return prefix + variables + functions + suffix


def process_classdef(node, indent):
    """Process class declaration.

    Args:
     - `node` -- `parso.python.tree.Class`, class AST
     - `indent` -- `int`, indentation tab size

    Returns:
     - `str` -- processed source string

    """
    prefix = ''
    suffix = ''

    ctx = ConvertContext(node, indent)

    children = iter(node.children)
    for child in children:
        prefix += walk(child, ctx)
        if child.type == 'operator' and child.value == ':':
            prefix += ctx.linesep
            break

    for child in children:
        bufpre, bufsuf = check_suffix(walk(child, ctx))
        prefix += bufpre
        suffix += bufsuf

    return ctx.finalize(prefix, suffix)


def process_funcdef(node, indent, *, async_ctx=None):
    """Process function declaration.

    Args:
     - `node` -- `parso.python.tree.Function`, function AST
     - `indent` -- `int`, indentation tab size

    Kwds:
     - `async_ctx` -- `parso.python.tree.Keyword`, `async` keyword AST node

    Returns:
     - `str` -- processed source string

    """
    if async_ctx is None:
        prefix = ''
        ctx = ConvertContext(node, indent)
    else:
        prefix = async_ctx.get_code()
        ctx = ConvertContext(async_ctx, indent)
    suffix = ''

    children = iter(node.children)
    for child in children:
        prefix += walk(child, ctx)
        if child.type == 'operator' and child.value == ':':
            prefix += ctx.linesep
            break

    for child in children:
        bufpre, bufsuf = check_suffix(walk(child, ctx))
        prefix += bufpre
        suffix += bufsuf

    return ctx.finalize(prefix, suffix)


def walk_expr(node):
    """Walk expression part of variable assignment.

    Args:
     - `node` -- `Union[parso.python.tree.PythonLeaf, parso.python.tree.PythonNode]`, expression AST

    Returns:
     - `str` -- processed source string

    """


def process_expr(node):
    """Process variable assignment.

    Args:
     - `node` -- `parso.python.tree.ExprStmt`, assignment AST

    Returns:
     - `str` -- processed source string

    """
    string = ''

    # <Name: ...>
    string += node.children[0].get_code()
    # <Operator: =>
    string += node.children[0].get_code()
    # PythonLeaf / PythonNode<atom>
    string += walk_expr(node.children[0])

    return string


def has_walrus(node):
    """Check if node has assignment expression.

    Args:
     - `node` -- `Union[parso.python.tree.PythonNode, parso.python.tree.PythonLeaf]`, node to search

    Return:
     - `bool` -- if node has assignment expression

    """
    if node.type == 'test_namedexpr':
        return True
    if not hasattr(node, 'children'):
        return False
    return any(map(has_walrus, node.children))


def find_walrus(node, root=0):
    """Find node to insert walrus context.

    Args:
     - `node` -- `parso.python.tree.Module`, parso AST
     - `root` -- `int`, index for insertion (based on `node`)

    Returns:
     - `int` -- index for insertion (based on `node`)

    """
    for index, child in enumerate(node.children, start=1):
        if has_walrus(child):
            return root
        if child.get_first_leaf().column == 0:
            root = index
    return -1


def check_indent(node):
    """Check indentation tab size.

    Args:
     - `node` -- `parso.python.tree.Module`, parso AST

    Returns:
     - `int` -- indentation tab size

    """
    for child in node.children:
        if child.type != 'suite':
            if hasattr(child, 'children'):
                tabsize = check_indent(child)
                if tabsize > 0:
                    return tabsize
            continue
        return child.children[1].get_first_leaf().column
    return 0


def check_suffix(string):
    """Strip comments from string.

    Args:
     - `string` -- `str`, buffer string

    Returns:
     - `str` -- prefix comments
     - `str` -- suffix strings

    """
    prefix = ''
    suffix = ''

    lines = iter(string.splitlines(True))
    for line in lines:
        if line.strip().startswith('#'):
            prefix += line
            continue
        suffix += line
        break

    for line in lines:
        suffix += line
    return prefix, suffix


def process_module(node):
    """Walk top nodes of the AST module.

    Args:
     - `node` -- `parso.python.tree.Module`, parso AST

    Envs:
     - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)

    Returns:
     - `str` -- processed source string

    """
    WALRUS_TABSIZE = int(os.getenv('WALRUS_TABSIZE', __walrus_tabsize__))

    prefix = ''
    suffix = ''

    tab = check_indent(node) or WALRUS_TABSIZE
    ctx = ConvertContext(node, tabsize=tab)
    pos = find_walrus(node)

    for index, child in enumerate(node.children):
        if index < pos:
            prefix += walk(child, ctx)
        else:
            bufpre, bufsuf = check_suffix(walk(child, ctx))
            prefix += bufpre
            suffix += bufsuf
    return ctx.finalize(prefix, suffix)


def guess_linesep(node):  # pylint: disable=inconsistent-return-statements
    """Guess line separator based on source code.

    Args:
     - `node` -- `Union[parso.python.tree.Module, parso.python.tree.PythonNode, parso.python.tree.PythonLeaf]`,
                 parso AST

    Returns:
     - `str` -- line separator

    """
    root = node.get_root_node()
    code = root.get_code()

    pool = {
        '\r': 0,
        '\r\n': 0,
        '\n': 0,
    }
    for line in code.splitlines(True):
        if line.endswith('\r'):
            pool['\r'] += 1
        elif line.endswith('\r\n'):
            pool['\r\n'] += 1
        else:
            pool['\n'] += 1

    sort = sorted(pool, key=lambda k: pool[k])
    if pool[sort[0]] == pool[sort[1]]:
        return get_linesep()
    return sort[0]


def get_linesep():
    """Get current line separator configuration.

    Returns:
     - `str` -- line separator

    """
    env = os.getenv('POSEUR_LINESEP', os.linesep)
    env_name = env.upper()
    if env_name == 'CR':
        return '\r'
    if env_name == 'CRLF':
        return '\r\n'
    if env_name == 'LF':
        return '\n'
    if env in ['\r', '\r\n', '\n']:
        return env
    raise EnvironError('invlid line separator %r' % env)


def walk(node, ctx=None):
    """Walk parso AST.

    Args:
     - `node` -- `Union[parso.python.tree.Module, parso.python.tree.PythonNode, parso.python.tree.PythonLeaf]`,
                 parso AST
     - `ctx` -- `ConvertContext`, conversion context

    Envs:
     - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)

    Returns:
     - `str` -- converted string

    """
    if isinstance(node, parso.python.tree.Module):
        return process_module(node)
    if ctx is None:
        raise ContextError('missing conversion context for node %r' % node)

    # string buffer
    string = ''

    if node.type == 'try_stmt':
        pass

    # if node.type in ('funcdef', 'classdef', 'if_stmt', 'while_stmt', 'for_stmt', 'with_stmt'):
    if node.type == 'funcdef':
        return process_funcdef(node, ctx.tabsize)

    if node.type == 'async_stmt':
        child_1st = node.children[0]
        child_2nd = node.children[1]

        flag_1st = child_1st.type == 'keyword' and child_1st.value == 'async'
        flag_2nd = child_2nd.type
        # flag_2nd = child_2nd.type in ('funcdef', 'with_stmt', 'for_stmt')

        if flag_1st and flag_2nd == 'funcdef':  # pragma: no cover
            return process_funcdef(child_2nd, ctx.tabsize, async_ctx=child_1st)

    if isinstance(node, parso.python.tree.PythonLeaf):
        string += node.get_code()

    if hasattr(node, 'children'):
        for child in node.children:
            if child.type == 'namedexpr_test':
                string += ctx.extract(child)
            else:
                string += walk(child, ctx)

    return string


def convert(string, source='<unknown>'):
    """The main conversion process.

    Args:
     - `string` -- `str`, context to be converted
     - `source` -- `str`, source of the context

    Envs:
     - `WALRUS_VERSION` -- convert against Python version (same as `--python` option in CLI)
     - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)

    Returns:
     - `str` -- converted string

    """
    # parse source string
    module = parse(string, source)

    # convert source string
    string = walk(module)

    # return converted string
    return string


def walrus(filename):
    """Wrapper works for conversion.

    Args:
     - `filename` -- `str`, file to be converted

    Envs:
     - `WALRUS_QUIET` -- run in quiet mode (same as `--quiet` option in CLI)
     - `WALRUS_ENCODING` -- encoding to open source files (same as `--encoding` option in CLI)
     - `WALRUS_VERSION` -- convert against Python version (same as `--python` option in CLI)
     - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)

    """
    WALRUS_QUIET = BOOLEAN_STATES.get(os.getenv('WALRUS_QUIET', '0').casefold(), False)
    if not WALRUS_QUIET:  # pragma: no cover
        print('Now converting %r...' % filename)

    # fetch encoding
    encoding = os.getenv('WALRUS_ENCODING', LOCALE_ENCODING)

    # file content
    with open(filename, 'r', encoding=encoding) as file:
        text = file.read()

    # do the dirty things
    text = convert(text, filename)

    # dump back to the file
    with open(filename, 'w', encoding=encoding) as file:
        file.write(text)


###############################################################################
# CLI & entry point

# default values
__cwd__ = os.getcwd()
__archive__ = os.path.join(__cwd__, 'archive')
__walrus_version__ = os.getenv('WALRUS_VERSION', WALRUS_VERSION[-1])
__walrus_encoding__ = os.getenv('WALRUS_ENCODING', LOCALE_ENCODING)
__walrus_linesep__ = os.getenv('WALRUS_LINESEP', os.linesep)
__walrus_tabsize__ = os.getenv('WALRUS_TABSIZE', '4')


def get_parser():
    """Generate CLI parser.

    Returns:
     - `argparse.ArgumentParser` -- CLI parser for walrus

    """
    parser = argparse.ArgumentParser(prog='walrus',
                                     usage='walrus [options] <python source files and folders...>',
                                     description='Back-port compiler for Python 3.8 assignment expressions.')
    parser.add_argument('-V', '--version', action='version', version=__version__)
    parser.add_argument('-q', '--quiet', action='store_true',
                        help='run in quiet mode')

    archive_group = parser.add_argument_group(title='archive options',
                                              description="duplicate original files in case there's any issue")
    archive_group.add_argument('-na', '--no-archive', action='store_false', dest='archive',
                               help='do not archive original files')
    archive_group.add_argument('-p', '--archive-path', action='store', default=__archive__, metavar='PATH',
                               help='path to archive original files (%s)' % __archive__)

    convert_group = parser.add_argument_group(title='convert options',
                                              description='compatibility configuration for none-unicode files')
    convert_group.add_argument('-c', '--encoding', action='store', default=__walrus_encoding__, metavar='CODING',
                               help='encoding to open source files (%s)' % __walrus_encoding__)
    convert_group.add_argument('-v', '--python', action='store', metavar='VERSION',
                               default=__walrus_version__, choices=WALRUS_VERSION,
                               help='convert against Python version (%s)' % __walrus_version__)
    convert_group.add_argument('-s', '--linesep', action='store', default=__walrus_linesep__, metavar='SEP',
                               help='line separator to process source files (%r)' % __walrus_linesep__)
    convert_group.add_argument('-nl', '--no-linting', action='store_false', dest='linting',
                               help='do not lint converted codes')
    convert_group.add_argument('-t', '--tabsize', action='store', default=__walrus_tabsize__, metavar='INDENT',
                               help='indentation tab size (%s)' % __walrus_tabsize__, type=int)

    parser.add_argument('file', nargs='+', metavar='SOURCE', default=__cwd__,
                        help='python source files and folders to be converted (%s)' % __cwd__)

    return parser


def find(root):  # pragma: no cover
    """Recursively find all files under root.

    Args:
     - `root` -- `os.PathLike`, root path to search

    Returns:
     - `Generator[str, None, None]` -- yield all files under the root path

    """
    file_list = list()
    for entry in os.scandir(root):
        if entry.is_dir():
            file_list.extend(find(entry.path))
        elif entry.is_file():
            file_list.append(entry.path)
        elif entry.is_symlink():  # exclude symbolic links
            continue
    yield from file_list


def rename(path, root):
    """Rename file for archiving.

    Args:
     - `path` -- `os.PathLike`, file to rename
     - `root` -- `os.PathLike`, archive path

    Returns:
     - `str` -- the archiving path

    """
    stem, ext = os.path.splitext(path)
    name = '%s-%s%s' % (stem, uuid.uuid4(), ext)
    return os.path.join(root, name)


def main(argv=None):
    """Entry point for walrus.

    Args:
     - `argv` -- `List[str]`, CLI arguments (default: None)

    Envs:
     - `WALRUS_QUIET` -- run in quiet mode (same as `--quiet` option in CLI)
     - `WALRUS_ENCODING` -- encoding to open source files (same as `--encoding` option in CLI)
     - `WALRUS_VERSION` -- convert against Python version (same as `--python` option in CLI)
     - `WALRUS_LINESEP` -- line separator to process source files (same as `--linesep` option in CLI)
     - `WALRUS_LINTING` -- lint converted codes (same as `--linting` option in CLI)
     - `WALRUS_TABSIZE` -- indentation tab size (same as `--tabsize` option in CLI)

    """
    parser = get_parser()
    args = parser.parse_args(argv)

    # set up variables
    ARCHIVE = args.archive_path
    os.environ['WALRUS_VERSION'] = args.python
    os.environ['WALRUS_ENCODING'] = args.encoding
    os.environ['WALRUS_TABSIZE'] = str(args.tabsize)
    WALRUS_QUIET = os.getenv('WALRUS_QUIET')
    os.environ['WALRUS_QUIET'] = '1' if args.quiet else ('0' if WALRUS_QUIET is None else WALRUS_QUIET)
    WALRUS_LINTING = os.getenv('WALRUS_LINTING')
    os.environ['WALRUS_LINTING'] = '1' if args.linting else ('0' if WALRUS_LINTING is None else WALRUS_LINTING)

    linesep = args.linesep.upper()
    if linesep == 'CR':
        os.environ['POSEUR_LINESEP'] = '\r'
    elif linesep == 'CRLF':
        os.environ['POSEUR_LINESEP'] = '\r\n'
    elif linesep == 'LF':
        os.environ['POSEUR_LINESEP'] = '\n'
    elif args.linesep in ['\r', '\r\n', '\n']:
        os.environ['POSEUR_LINESEP'] = args.linesep
    else:
        raise EnvironError('invalid line separator %r' % args.linesep)

    # make archive directory
    if args.archive:  # pragma: no cover
        os.makedirs(ARCHIVE, exist_ok=True)

    # fetch file list
    filelist = list()
    for path in args.file:
        if os.path.isfile(path):
            if args.archive:  # pragma: no cover
                dest = rename(path, root=ARCHIVE)
                os.makedirs(os.path.dirname(dest), exist_ok=True)
                shutil.copy(path, dest)
            filelist.append(path)
        if os.path.isdir(path):  # pragma: no cover
            if args.archive:
                shutil.copytree(path, rename(path, root=ARCHIVE))
            filelist.extend(find(path))

    # check if file is Python source code
    def ispy(file): return (os.path.isfile(file) and (os.path.splitext(file)[1] in ('.py', '.pyw')))
    filelist = sorted(filter(ispy, filelist))

    # if no file supplied
    if not filelist:  # pragma: no cover
        parser.error('argument PATH: no valid source file found')

    # process files
    if mp is None or CPU_CNT <= 1:
        [walrus(filename) for filename in filelist]  # pylint: disable=expression-not-assigned # pragma: no cover
    else:
        with mp.Pool(processes=CPU_CNT) as pool:
            pool.map(walrus, filelist)


if __name__ == '__main__':
    sys.exit(main())
